### PR TITLE
Close IOStreams

### PR DESCRIPTION
**Pull request by bio-boris**

### CODE REVIEW

File: src/us/kbase/common/service/JsonClientCaller.java [hunk 1]
```diff
@@ -357,108 +357,127 @@ public static AuthToken requestTokenFromKBase(
 			conn.setFixedLengthStreamingMode(size);
 		}
 		// Write real data into http output stream
-		writeRequestDataStatic(method, arg, conn.getOutputStream(), id, context);
+
+		try {
+			writeRequestDataStatic(method, arg, conn.getOutputStream(), id, context);
```
Comment (bio-boris) [inline src/us/kbase/common/service/JsonClientCaller.java]: wrap here

File: src/us/kbase/common/service/JsonClientCaller.java [hunk 2]
```diff
@@ -357,108 +357,127 @@ public static AuthToken requestTokenFromKBase(
...
+						try {
+							fos.close();
+						} catch (Exception ignore) {
+						}
+				}
 			}
 		}
+		finally {
+			if(istream !=null)
+				istream.close();
```
Comment (bio-boris) [inline src/us/kbase/common/service/JsonClientCaller.java]: wrap here

File: src/us/kbase/common/service/JsonClientCaller.java [hunk 3]
```diff
@@ -357,108 +357,127 @@ public static AuthToken requestTokenFromKBase(
...
-				throw new ServerException(retError.get("message"),
-						new Integer(retError.get("code")), retError.get("name"),
-						data);
-			}
-			if (res == null && ret) {
-				throw new ServerException("An unknown server error occured",
-						0, "Unknown", null);
+		InputStream istream = null;
+
+		try {
```
Comment (bio-boris) [inline src/us/kbase/common/service/JsonClientCaller.java]: wrap here

File: src/us/kbase/common/service/JsonClientCaller.java (hunk 1 above)
Comment (MrCreosote) [inline src/us/kbase/common/service/JsonClientCaller.java]: I'd probably do
```
try (final OutputStream out = conn.getOutputStream()) {
    writeRequestDataStatic(method, arg, out, id, context);
}
```
A little more compact and lets the language handle the close.

File: src/us/kbase/common/service/JsonClientCaller.java (hunk 3 above)
Comment (MrCreosote) [inline src/us/kbase/common/service/JsonClientCaller.java]: I guess you can't really use try with resources here

**Comment by MrCreosote** [comment]
I don't think there's much in the way of tests for this in this repo, unfortunately.

File: src/us/kbase/common/service/JsonClientCaller.java [hunk 4]
```diff
@@ -357,108 +357,125 @@ public static AuthToken requestTokenFromKBase(
 			conn.setFixedLengthStreamingMode(size);
 		}
 		// Write real data into http output stream
-		writeRequestDataStatic(method, arg, conn.getOutputStream(), id, context);
+		try (final OutputStream out = conn.getOutputStream()) {
+			writeRequestDataStatic(method, arg, out, id, context);
+		}
+		catch (IOException ignore){
```
Comment (MrCreosote) [inline src/us/kbase/common/service/JsonClientCaller.java]: I'm not a fan of ignoring exceptions - is there a reason not to pass it up the stack? Presumably this shouldn't happen unless the stream errors on a close, which indicates a problem.

File: src/us/kbase/common/service/JsonClientCaller.java [hunk 5]
```diff
@@ -357,108 +357,126 @@ public static AuthToken requestTokenFromKBase(
 			conn.setFixedLengthStreamingMode(size);
 		}
 		// Write real data into http output stream
-		writeRequestDataStatic(method, arg, conn.getOutputStream(), id, context);
+		try (final OutputStream out = conn.getOutputStream()) {
+			writeRequestDataStatic(method, arg, out, id, context);
+		}
+		catch (IOException e){
```
Comment (MrCreosote) [inline src/us/kbase/common/service/JsonClientCaller.java]: This is a no-op - the whole block can just be removed.

**Comment by MrCreosote** [comment]
Only question I have left:

> did you figure out what the deal was with the wrapped input stream?

**Comment by bio-boris** [comment]
Based on https://www.google.com/search?q=UnclosableInputStream 
An input stream that is ignoring all request to close():
The point is to trap close() calls so that the underlying stream is not closed. Used for methods that automatically close a stream at EOF, even though it is undesirable for the stream to be closed

**Comment by MrCreosote** [comment]
UnclosableInputStream in this case is defined in the file though, so it may not have the same behavior these other implementations have.
Why is it bad to close the stream on an EOF?

**Comment by MrCreosote** [comment]
Also, confusingly, it looks like it doesn't ignore close calls: https://github.com/kbase/java_common/blob/develop/src/us/kbase/common/service/JsonClientCaller.java#L567

File: src/us/kbase/common/service/JsonClientCaller.java [hunk 6]
```diff
@@ -565,7 +586,19 @@ public int available() throws IOException {
 		
 		@Override
 		public void close() throws IOException {
-			isClosed = true;
+			try{
```
Comment (bio-boris) [inline src/us/kbase/common/service/JsonClientCaller.java]: new

**Comment by MrCreosote** [comment]
Ok, if this fixes the problem we'll add release notes and make a release

**Pull request merged**